utils/data_uri.go

```diff
@@ -0,0 +1,64 @@
+// Copyright 2023 The kubecfg authors
+//
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package utils
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+)
+
+func ToDataURL(code string) string {
+	return fmt.Sprintf("data:,%s", url.PathEscape(code))
+}
+
+type dataURIImporter struct {
+}
+
+func newDataURIImporter() *dataURIImporter {
+	return &dataURIImporter{}
+}
+
+func (o *dataURIImporter) RoundTrip(req *http.Request) (*http.Response, error) {
+	content, _, err := expandDataURL(req.URL.String())
+	if err != nil {
+		return nil, err
+	}
+
+	return simpleHTTPResponse(req, http.StatusOK, io.NopCloser(strings.NewReader(content))), nil
+}
+
+func expandDataURL(pathURL string) (string, string, error) {
+	content, err := url.PathUnescape(strings.TrimPrefix(pathURL, "data:,"))
+	if err != nil {
+		return "", "", err
+	}
+	cwd, err := os.Getwd()
+	if err != nil {
+		return "", "", err
+	}
+
+	foundAt, err := PathToURL(cwd)
```

Comment:
I know this is what the original code had, but I think this is incorrect, and could cause the cached data URL "contents" to collide with other data URLs.

I think `foundAt` should be the original data URL.

Reply:
yeah it's complicated.

The `foundAt` is used as the base to resolve relative paths when the data uri content itself then contains imports.

The main use case for data urls is to allow users to pass inline code to flags like in:

```console
$ ./kubecfg show foo.yaml --overlay-code '{name:: (importstr "README.md")[2:9]}'
---
apiVersion: v1
kind: Foo
spec:
  name: kubecfg
```

Data uris here are just an implementation hack. We could have implemented this `--overlay-code` functionality differently. But it was quite natural: all commands like show, update, validate, diff etc know how to read a url, so all we had to do is to turn "some_url" "--overlay-code" "some_str" into a `data:,(import "some_url") + (some_str)` (proper escaping not shown here) and feed this new url into the same existing logic that would process the input url.

Since the main use case was to use data uris internally to handle CLI parameters, using `cwd` was natural, since the user's intent was captured: it was reasonable that relative paths present in the literal string passed to the command were resolved relative to the current directory.

However, here we're expanding the ability to use this data urls also inside other files, and so the relative path resolution should be relative to those files, like any other import.

So I guess I have to find a way to weave this context through

Reply:
just to be clear, if I set `foundAt = pathURL.String()` my example command above will return this error:

```
RUNTIME ERROR: Get "data:///README.md": invalid data url: missing ','
```

Reply:
but since we're using `foundAt` as a cache key in:

```
		if c, ok := importer.cache[foundAt]; ok {
			return c, foundAt, nil
		}
```

we clearly cannot return whatever was the `importedFrom` who imported the data url, as the new `foundAt`, we need to return something that uniquely captures the data url **and** also the location where relative paths should be resolved against.

Reply:
I see a couple of options:

1. using a custom media type parameter attribute name, like `data:;kubecfg_imported_from=someuriencodeduri,somedata`

2. using a custom uri scheme like `kubecfgdata:///?base=someuriencodeduri&data=someuriencodeddata`